packages/react-next/src/components/Callout/CalloutContent.base.tsx

```diff
@@ -50,136 +51,353 @@ const OFF_SCREEN_STYLE = { opacity: 0, filter: 'opacity(0)' };
 // calloutMain (the Popup component within the callout)
 const ARIA_ROLE_ATTRIBUTES = ['role', 'aria-roledescription'];
 
-export interface ICalloutState {
-  positions?: ICalloutPositionedInfo;
-  slideDirectionalClassName?: string;
-  calloutElementRect?: ClientRect;
-  heightOffset?: number;
+const DEFAULT_PROPS = {
+  preventDismissOnLostFocus: false,
+  preventDismissOnScroll: false,
+  preventDismissOnResize: false,
+  isBeakVisible: true,
+  beakWidth: 16,
+  gapSpace: 0,
+  minPagePadding: 8,
+  directionalHint: DirectionalHint.bottomAutoEdge,
+} as const;
+
+function useTargets({ target }: ICalloutProps, calloutElement: React.RefObject<HTMLDivElement | null>) {
+  const previousTargetProp = React.useRef<
+    Element | string | MouseEvent | Point | React.RefObject<Element> | null | undefined
+  >();
+
+  const targetRef = React.useRef<Element | MouseEvent | Point | null>(null);
+  /**
+   * Stores an instance of Window, used to check
+   * for server side rendering and if focus was lost.
+   */
+  const targetWindowRef = React.useRef<Window>();
+
+  // If the target element changed, find the new one. If we are tracking
+  // target with class name, always find element because we do not know if
+  // fabric has rendered a new element and disposed the old element.
+  if (!target || target !== previousTargetProp.current || typeof target === 'string') {
+    const currentElement = calloutElement.current;
+    if (target) {
+      if (typeof target === 'string') {
+        const currentDoc: Document = getDocument(currentElement)!;
+        targetRef.current = currentDoc ? currentDoc.querySelector(target) : null;
+        targetWindowRef.current = getWindow(currentElement)!;
+      } else if ('stopPropagation' in target) {
+        targetWindowRef.current = getWindow(target.target as Element)!;
+        targetRef.current = target;
+      } else if ('getBoundingClientRect' in target) {
+        targetWindowRef.current = getWindow(currentElement)!;
+        targetRef.current = target;
+      } else if ('current' in target && target.current !== undefined) {
+        targetRef.current = target.current;
+        targetWindowRef.current = getWindow(target.current)!;
+      } else {
+        targetWindowRef.current = getWindow(currentElement)!;
+        targetRef.current = target as Point;
+      }
+    } else {
+      targetWindowRef.current = getWindow(currentElement)!;
+    }
+    previousTargetProp.current = target;
+  }
+
+  return [targetRef, targetWindowRef] as const;
```

Comment:
curious why you need windowref.

Reply:
To cache the window reference used by other functions, rather than re-fetching it every render.